packages/backend/src/history/historyNotifier.ts

```diff
@@ -0,0 +1,48 @@
+/**********************************************************************
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ***********************************************************************/
+
+import { type Disposable, type Webview } from '@podman-desktop/api';
+import * as podmanDesktopApi from '@podman-desktop/api';
+import path from 'node:path';
+import { Messages } from '/@shared/src/messages/Messages';
+import { BOOTC_HISTORY_FILENAME } from '/@shared/src/models/bootc';
+
+export class HistoryNotifier implements Disposable {
+  #watcher: podmanDesktopApi.FileSystemWatcher;
+
+  constructor(
+    private webview: Webview,
+    private readonly storagePath: string,
+  ) {
+    this.#watcher = podmanDesktopApi.fs.createFileSystemWatcher(path.join(this.storagePath, BOOTC_HISTORY_FILENAME));
+    this.#watcher.onDidChange(this.notify.bind(this));
+    this.#watcher.onDidCreate(this.notify.bind(this));
+    this.#watcher.onDidDelete(this.notify.bind(this));
+  }
+
+  async notify(): Promise<void> {
+    await this.webview.postMessage({
```

Comment:
This creates an odd circular dependency where a change to the file keeps the cached data in backend but triggers an event to the frontend, which calls back to listHistoryInfo() to get the info, which triggers the actual file reload. The file is reloaded on any call from the frontend, even if it hasn't changed.

I know this is unlikely to cause issues in reality, but it seems like this function should be the one triggering a reload (which wouldn't be great b/c it could reload lots even if we're not using the history atm), or better yet just clearing the history and having history.getHistory() automatically reload when there's no data. i.e. history.getHistory() is a little smarter and the current forced reload in listHistoryInfo() could be removed.

Reply:
Needed in order to get RPCReadable working between frontend and backend. The notifier allows the frontend to check / pull the newest one, rather than using the backend. I tried implementing it the opposite way (like you had mentioned).. but ran into the opposite which was caching issues.

This is similar to the aistudio implementation. and is intentional to not post in values and used to just notify the frontend, we end up retrieving the file with listHistoryInfo like what you said in here:

```
export const historyInfo: Readable<BootcBuildInfo[]> = RPCReadable<BootcBuildInfo[]>(
  [],
  [Messages.MSG_HISTORY_UPDATE],
  bootcClient.listHistoryInfo,
);
```

With stores, the only time that it is actually reloaded is the initial loading (when switching to webview) or the notify from postMessage.

Reply:
I'm not suggesting posting values, just that I don't like the pattern of telling the frontend that there is a change but keeping cached values until frontend asks you for data (and forced reload either way). One path watching for file system changes and forcing a reload either way, but others like quickpick using 'cached' data.

I am fine approving as-is since I can't find any path that would fail today except if you touch the file directly, but I think we need to track the inconsistency since it could lead to odd failures in the future if you're not aware of it.